### PR TITLE
common/admin_socket: validate command json before feeding it to hook

### PR DESCRIPTION
**Pull request by tchaikov**
Signed-off-by: Kefu Chai <kchai@redhat.com>

### CODE REVIEW

**Comment by tchaikov** [comment]
i tested by applying #20274 on top of this PR. and ceph-mon survived `bin/ceph_test_admin_socket_output --mon /tmp/user/1000/ceph-asok.*`

File: src/common/admin_socket.cc [hunk 1]
```diff
@@ -439,6 +440,166 @@ bool AdminSocket::do_accept()
   return rval;
 }
 
+namespace {
+template <typename Func>
+bool find_first_in(boost::string_view s, const char *delims, Func&& f)
```
Comment (jcsp) [inline src/common/admin_socket.cc]: I think we can use std::string_view now that it's standardized in C++17

Comment (tchaikov) [reply src/common/admin_socket.cc]: @jcsp yes, but https://github.com/ceph/ceph/blob/master/src/include/str_list.h#L16 is still using its boost counterpart. i will do the switch over in another PR.

File: src/common/admin_socket.cc [hunk 2]
```diff
@@ -439,6 +440,166 @@ bool AdminSocket::do_accept()
...
+
+bool validate_arg_type(const boost::string_view value,
+		       const boost::string_view type,
+		       const arg_desc_t& desc,
+		       std::ostream& os)
+{
+  if (type == "CephString") {
+    return true;
+  } else if (type == "CephIPAddr") {
+    os << "failed to parse addr '" << value << "', should be ip:[port]";
```
Comment (jcsp) [inline src/common/admin_socket.cc]: Missing the actual validation for this type of value?

**Comment by jcsp** [comment]
I wonder if this code belongs in the cmdparse file instead of admin socket, since I imagine we'll want to re-use it for MCommand commands at some point too?

**Comment by tchaikov** [comment]
@jcsp  comments addressed and repushed.

**Comment by jcsp** [comment]
Looks good!  I have wished we had this validation server side for a long time :tada:

**Comment by tchaikov** [comment]
retest this please

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-02-15_15:55:27-rados-wip-kefu-testing-2018-02-15-0045-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2018-02-16_01:22:51-rados-wip-kefu-testing-2018-02-16-0353-distro-basic-mira/

**Pull request merged**